### PR TITLE
JaxSolver fails when using GPU support with no input parameters

### LINKED ISSUE #3422: [Bug]: JaxSolver fails when using GPU support with no input parameters
### PyBaMM Version

23.5

### Python Version

3.10.12

### Describe the bug

When specified with input parameters, the JaxSolver will parallelise the solve across parameter sets either using `asyncio` (cpu) or jax-`vmap` (gpu). When specified without input parameters the cpu pathway continues to solve the model correctly; however, when the solver is called without an input parameter list in a gpu-enabled jax environment then the solver fails. This is because the `vmap` function requires at least one input argument to contain a non-empty array.

### Steps to Reproduce

```python
import pybamm
import numpy as np
import matplotlib.pyplot as plt

import jax

platform = jax.lib.xla_bridge.get_backend().platform.casefold()
print("Platform: ", platform)

model = pybamm.BaseModel()

x = pybamm.Variable("x")
y = pybamm.Variable("y")

dxdt = 4 * x - 2 * y
dydt = 3 * x - y

model.convert_to_format = 'jax'
model.rhs = {x: dxdt, y: dydt}
model.initial_conditions = {x: pybamm.Scalar(1), y: pybamm.Scalar(2)}
model.variables = {"x": x, "y": y, "z": x + 4 * y}
disc = pybamm.Discretisation()  # use the default discretisation
disc.process_model(model);

solver = pybamm.JaxSolver()
t = np.linspace(0, 1, 20)
solution = solver.solve(model, t)

t_sol, y_sol = solution.t, solution.y  # get solution times and states
x = solution["x"]  # extract and process x from the solution
y = solution["y"]  # extract and process y from the solution

print(x.data)
print(y.data)
```

### Relevant log output

```shell
Platform:  gpu
jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jsb/repos/jsbrittain/PyBaMM/test.py", line 27, in <module>
    solution = solver.solve(model, t)
  File "/home/jsb/repos/jsbrittain/PyBaMM/pybamm/solvers/base_solver.py", line 853, in solve
    new_solution = self._integrate(
  File "/home/jsb/repos/jsbrittain/PyBaMM/pybamm/solvers/jax_solver.py", line 240, in _integrate
    y.extend(jax.vmap(self._cached_solves[model])(inputs_v))
ValueError: vmap wrapped function must be passed at least one argument containing an array, got empty *args=({},) and **kwargs={}
```

### PR DESCRIPTION
**Pull request by jsbrittain**
# Description

When specified with input parameters, the JaxSolver will parallelise the solve across parameter sets either using asyncio (cpu) or jax-vmap (gpu). When specified without input parameters the cpu pathway continues to solve the model correctly; however, when the solver is called without an input parameter list in a gpu-enabled jax environment then the solver fails. This is because the vmap function requires at least one input argument to contain a non-empty array.

A fix is to ensure that input parameter sets of length 0 or 1 are directed towards the cpu pathway, since this parallelisation occurs mostly over parameter sets, rather than within solves.

Fixes #3422 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [X] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [X] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [X] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [X] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

** Existing tests should cover this scenario, but depend upon GPU runners which are work-in-progress. **

### CODE REVIEW

**Review by brosaplanella — APPROVED** [review]
Looks good, thanks! Let's hold merging this until the wheel publishing is fixed.

**Comment by agriyakhetarpal** [comment]
Thanks for this @jsbrittain, I tested this on Windows locally and I can confirm that #3371 is fixed through the changes in this PR. I didn't know it would be as easy as this but I get it because of my general inexperience with Jax. Could you please close that issue too here and add it to the CHANGELOG?

I have opened a new issue about bumping the `jax` and `jaxlib` versions so that GPU support can be targeted across platforms and for other reasons, see #3443

**Review by Saransh-cpp — COMMENTED** [review]
The wheel and rc situation is fixed now. The CHANGELOG line should be moved to the unreleased section and everything should be good to go.

**Comment by agriyakhetarpal** [comment]
The [failing doctests](https://github.com/pybamm-team/PyBaMM/actions/runs/6638247118/job/18034727563#step:8:453) here should not be a worry. I would suggest setting that configuration value to `True`, though.

**Comment by agriyakhetarpal** [comment]
The [failing doctests](https://github.com/pybamm-team/PyBaMM/actions/runs/6638247118/job/18034727563#step:8:453) here should not be a worry. I would suggest setting that configuration value to `False`, though.

**Comment by agriyakhetarpal** [comment]
The example notebooks issue should be this one: #3415, which came after the changes made in #3198.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # [Unreleased](https://github.com/pybamm-team/PyBaMM/)
 
+- Fixed a bug where the JaxSolver would fails when using GPU support with no input parameters ([#3423](https://github.com/pybamm-team/PyBaMM/pull/3423))
```
Comment (Saransh-cpp) [inline CHANGELOG.md:3]:
```suggestion
## Bug fixes

- Fixed a bug where the JaxSolver would fails when using GPU support with no input parameters ([#3423](https://github.com/pybamm-team/PyBaMM/pull/3423))
```

**Review by Saransh-cpp — APPROVED** [review]
Thanks, @jsbrittain! Sorry for all the release mess 😬

**Comment by agriyakhetarpal** [comment]
@brosaplanella: I missed talking about it in the meeting—should this PR be included in the release given that this is a bug fix? I think #3443 ought to be taken a look at too otherwise the GPU support mentioned in our CHANGELOG for v23.9rc0 would not be available for users on some platforms (e.g., macOS with Metal requires `v0.4.11`—we are still on `v0.4.8`).

I would be happy to write a PR for that this week and incorporate @jsbrittain's suggestions on it into consideration, please let me know if that would be needed

**Comment by brosaplanella** [comment]
If we can include it that would be great. Tagging @Saransh-cpp so he is aware and can provide some input.

**Comment by Saransh-cpp** [comment]
I'll merge this and add it in the rc1 release.

**Pull request merged**